package/src/completionCacheManager/completionCacheManager.ts

```diff
@@ -0,0 +1,23 @@
+import * as monaco from 'monaco-editor/esm/vs/editor/editor.api';
+import * as ls from 'vscode-languageserver-types';
+
+export const createCompletionCacheManager = (getFromLanguageService: GetFromLanguageService) => {
+    let completionList: ls.CompletionList | undefined;
+    let lastWord: string | undefined;
+
+    return {
+        getCompletionItems: async (word: string | undefined, resource: monaco.Uri, position: ls.Position) => {
+            const isIncluded = word?.includes(lastWord);
+            const shouldGetItems = !lastWord || !word || !isIncluded;
+            if (shouldGetItems) {
+                completionList = await getFromLanguageService(resource, position);
+            }
+
+            lastWord = word;
```

Comment:
```suggestion
            lastWord = word;
            }

```

If we update the last work when the completion items haven't changed, don't we make it possible to not get completion items even when the new word make sense for the original `completionList`?

For example, if we call this with these words: "asdf", "a", "df". "df" won't hit the cache because "a" proceeded it, but would have if we omit "a".

Reply:
Let's simulate what happens when we call it with these words in this order (I am mimicking it character by character as the user would type it on their keyboard. Copy-pasting doesn't trigger the CompletionAdapter at all):

"a" - miss - calls LS (language service)
"as" - hit - doesn't call LS
"asd" - hit
"asdf" - hit
"asd" - CompletionAdapter isn't triggered when deleting
"as" - not triggered
"a" - not triggered
"" - not triggered
"a" - miss as lastWord is "asdf" and it is not contained in "a"
"" - not triggered
"d" - miss as "a" is not contained in "d"
"df" - hit as "d" is included in "df"
So, if I understood you correctly, I think we are okay in this case. What do you think?

Reply:
There's a few things I'm pretty sure that will happen that I think are pretty weird:

> "as" - hit - doesn't call LS
> "asd" - hit
> "asdf" - hit

Are we sure it makes sense to use cached results with these? I'd expect "asdf" to return pretty different results from "a"

> "a" - miss as lastWord is "asdf" and it is not contained in "a"

Even though we updated the last word, the last completion result _was_ "a". If we only updated the word when we update this completion result, this would be a cache hit.

Reply:
Are you expecting "asdf" to return completely different results or just a subset of the results returned from "a"? The Monaco editor filters the results, so as long as the results from "asdf" are a subset, we're fine. Previously, before setting incomplete: true, the Monaco editor would call provideCompletionItems only on the first character (and also on the fourth). I attempted to replicate this logic here (except for the fourth character as it produced strange results).